[Prototype] Introduce intermediate result storage for async search

fcofdez opened the pull request:
This is a first prototype to reduce the memory overhead for on-going searches. The approach taken is storing intermediate results in an index and perform incremental reductions that are also stored in an index. Shard searches are retried until they're successful (there's still some work to do around timeouts). I've added a new search mode `persistent search` that allowed me to implement this prototype independently from the rest of the search modes.

There's still quite a bit of work to do around error handling and improve certain parts of the code that are a bit hacky.

fcofdez committed:
Introduce persistent async search

fcofdez committed:
Minor cleanup

fcofdez committed:
Fix inconsistent comparable

jimczi reviewed (commented):
That's a great start. I left some high level comments but I think we should split the remaining work into smaller chunks. We'll need a meta issue somewhere to follow the progress but this PR is a huge step forward. 
My next action item is to focus the review on the storage and the new reduce design but I wanted to give this early feedback first! Nice work @fcofdez !

fcofdez committed:
Allow partial results by default

fcofdez committed:
Add can match phase to persistent search

fcofdez committed:
Merge remote-tracking branch 'origin/master' into persistent-async-search-proto

fcofdez committed:
Add partial results to an in memory cache

Include all options for the search request
Introduce circuit breaker during reductions

fcofdez committed:
Merge remote-tracking branch 'origin/master' into persistent-async-search-proto

fcofdez committed:
Fix test

fcofdez commented:
Closing this as we decided not to take this path.
